File: mycollab-scheduler/src/main/scala/com/mycollab/module/project/schedule/email/service/ProjectTaskRelayEmailNotificationActionImpl.scala
/**
 * This file is part of mycollab-scheduler.
 *
 * mycollab-scheduler is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-scheduler is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-scheduler.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.project.schedule.email.service

import com.hp.gagawa.java.elements.{Span, Text}
import com.mycollab.common.i18n.GenericI18Enum
import com.mycollab.common.{MonitorTypeConstants, NotificationType}
import com.mycollab.core.utils.StringUtils
import com.mycollab.html.{FormatUtils, LinkUtils}
import com.mycollab.module.mail.MailUtils
import com.mycollab.module.project.domain._
import com.mycollab.module.project.i18n.OptionI18nEnum.Priority
import com.mycollab.module.project.i18n.{MilestoneI18nEnum, OptionI18nEnum, TaskI18nEnum}
import com.mycollab.module.project.service._
import com.mycollab.module.project.{ProjectLinkGenerator, ProjectResources, ProjectTypeConstants}
import com.mycollab.module.user.AccountLinkGenerator
import com.mycollab.module.user.domain.SimpleUser
import com.mycollab.module.user.service.UserService
import com.mycollab.schedule.email.format.{DateFieldFormat, FieldFormat, I18nFieldFormat}
import com.mycollab.schedule.email.project.ProjectTaskRelayEmailNotificationAction
import com.mycollab.schedule.email.{ItemFieldMapper, MailContext}
import com.mycollab.spring.AppContextUtil
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
  * @author MyCollab Ltd.
  * @since 4.6.0
  */
@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class ProjectTaskRelayEmailNotificationActionImpl extends SendMailToFollowersAction[SimpleTask] with ProjectTaskRelayEmailNotificationAction {
  private val LOG = LoggerFactory.getLogger(classOf[ProjectTaskRelayEmailNotificationActionImpl])

  @Autowired var projectTaskService: ProjectTaskService = _

  @Autowired var projectNotificationService: ProjectNotificationSettingService = _

  private val mapper = new TaskFieldNameMapper

  protected def buildExtraTemplateVariables(context: MailContext[SimpleTask]) {
    val emailNotification = context.getEmailNotification

    val summary = "#" + bean.getTaskkey + " - " + bean.getName
    val summaryLink = ProjectLinkGenerator.generateTaskPreviewFullLink(siteUrl, bean.getTaskkey, bean.getProjectShortname)

    val avatarId = if (projectMember != null) projectMember.getMemberAvatarId else ""
    val userAvatar = LinkUtils.newAvatar(avatarId)

    val makeChangeUser = userAvatar.toString + " " + emailNotification.getChangeByUserFullName
    val actionEnum = emailNotification.getAction match {
      case MonitorTypeConstants.CREATE_ACTION => TaskI18nEnum.MAIL_CREATE_ITEM_HEADING
      case MonitorTypeConstants.UPDATE_ACTION => TaskI18nEnum.MAIL_UPDATE_ITEM_HEADING
      case MonitorTypeConstants.ADD_COMMENT_ACTION => TaskI18nEnum.MAIL_COMMENT_ITEM_HEADING
    }
  
    contentGenerator.putVariable("projectName", bean.getProjectName)
    contentGenerator.putVariable("projectNotificationUrl", ProjectLinkGenerator.generateProjectSettingFullLink(siteUrl, bean.getProjectid))
    contentGenerator.putVariable("actionHeading", context.getMessage(actionEnum, makeChangeUser))
    contentGenerator.putVariable("name", summary)
    contentGenerator.putVariable("summaryLink", summaryLink)
  }

  protected def getBeanInContext(notification: ProjectRelayEmailNotification): SimpleTask =
    projectTaskService.findById(notification.getTypeid.toInt, notification.getSaccountid)

  protected def getItemName: String = StringUtils.trim(bean.getName, 100)
  
  override protected def getProjectName: String = bean.getProjectName
  
  protected def getCreateSubject(context: MailContext[SimpleTask]): String = context.getMessage(TaskI18nEnum
    .MAIL_CREATE_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

  protected def getUpdateSubject(context: MailContext[SimpleTask]): String = context.getMessage(TaskI18nEnum
    .MAIL_UPDATE_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

  protected def getCommentSubject(context: MailContext[SimpleTask]): String = context.getMessage(TaskI18nEnum
    .MAIL_COMMENT_ITEM_SUBJECT, bean.getProjectName, context.getChangeByUserFullName, getItemName)

  protected def getItemFieldMapper: ItemFieldMapper = mapper

  protected def getListNotifyUsersWithFilter(notification: ProjectRelayEmailNotification): Set[SimpleUser] = {
    import scala.collection.JavaConverters._
    val notificationSettings = projectNotificationService.findNotifications(notification.getProjectId, notification.getSaccountid).asScala.toList
    var notifyUsers = notification.getNotifyUsers.asScala.toSet

    if (notificationSettings != null && notificationSettings.size > 0) {
      for (notificationSetting <- notificationSettings) {
        if (NotificationType.None.name == notificationSetting.getLevel) {
          notifyUsers = notifyUsers.filter(notifyUser => !(notifyUser.getUsername == notificationSetting.getUsername))
        }
        else if (NotificationType.Minimal.name == notificationSetting.getLevel) {
          val findResult = notifyUsers.find(notifyUser => notifyUser.getUsername == notificationSetting.getUsername);
          findResult match {
            case None => {
              val task = projectTaskService.findById(notification.getTypeid.toInt, notification.getSaccountid)
              if (notificationSetting.getUsername == task.getAssignuser) {
                val prjMember: SimpleUser = projectMemberService.getActiveUserOfProject(notificationSetting.getUsername,
                  notificationSetting.getProjectid, notificationSetting.getSaccountid)
                if (prjMember != null) {
                  notifyUsers += prjMember
                }
              }
            }
            case Some(user) => {}
          }
        }
        else if (NotificationType.Full.name == notificationSetting.getLevel) {
          val prjMember = projectMemberService.getActiveUserOfProject(notificationSetting.getUsername,
            notificationSetting.getProjectid, notificationSetting.getSaccountid)
          if (prjMember != null) {
            notifyUsers += prjMember
          }
        }
      }
    }

    notifyUsers
  }

  class TaskFieldNameMapper extends ItemFieldMapper {
    put(Task.Field.name, GenericI18Enum.FORM_NAME, isColSpan = true)
    put(Task.Field.startdate, new DateFieldFormat(Task.Field.startdate.name, GenericI18Enum.FORM_START_DATE))
    put(Task.Field.enddate, new DateFieldFormat(Task.Field.enddate.name, GenericI18Enum.FORM_END_DATE))
    put(Task.Field.duedate, new DateFieldFormat(Task.Field.duedate.name, GenericI18Enum.FORM_DUE_DATE))
    put(Task.Field.percentagecomplete, TaskI18nEnum.FORM_PERCENTAGE_COMPLETE)
    put(Task.Field.priority, new I18nFieldFormat(Task.Field.priority.name, GenericI18Enum.FORM_PRIORITY, classOf[Priority]))
    put(Task.Field.assignuser, new AssigneeFieldFormat(Task.Field.assignuser.name, GenericI18Enum.FORM_ASSIGNEE))
    put(Task.Field.isestimated, TaskI18nEnum.FORM_IS_ESTIMATED)
    put(Task.Field.remainestimate, TaskI18nEnum.FORM_REMAIN_ESTIMATE)
    put(Task.Field.milestoneid, new MilestoneFieldFormat(Task.Field.milestoneid.name, MilestoneI18nEnum.SINGLE))
    put(Task.Field.parenttaskid, new TaskFieldFormat(Task.Field.parenttaskid.name, TaskI18nEnum.FORM_PARENT_TASK))
    put(Task.Field.description, GenericI18Enum.FORM_DESCRIPTION)
    put(Task.Field.status, GenericI18Enum.FORM_STATUS)
  }

  class AssigneeFieldFormat(fieldName: String, displayName: Enum[_]) extends FieldFormat(fieldName, displayName) {

    def formatField(context: MailContext[_]): String = {
      val task = context.getWrappedBean.asInstanceOf[SimpleTask]
      if (task.getAssignuser != null) {
        val userAvatarLink = MailUtils.getAvatarLink(task.getAssignUserAvatarId, 16)
        val img = FormatUtils.newImg("avatar", userAvatarLink)
        val userLink = AccountLinkGenerator.generatePreviewFullUserLink(MailUtils.getSiteUrl(task.getSaccountid), task.getAssignuser)
        val link = FormatUtils.newA(userLink, task.getAssignUserFullName)
        FormatUtils.newLink(img, link).write
      }
      else {
        new Span().write
      }
    }

    def formatField(context: MailContext[_], value: String): String = {
      if (StringUtils.isBlank(value)) {
        new Span().write
      } else {
        val userService = AppContextUtil.getSpringBean(classOf[UserService])
        val user = userService.findUserByUserNameInAccount(value, context.getUser.getAccountId)
        if (user != null) {
          val userAvatarLink = MailUtils.getAvatarLink(user.getAvatarid, 16)
          val userLink = AccountLinkGenerator.generatePreviewFullUserLink(MailUtils.getSiteUrl(user.getAccountId), user.getUsername)
          val img = FormatUtils.newImg("avatar", userAvatarLink)
          val link = FormatUtils.newA(userLink, user.getDisplayName)
          FormatUtils.newLink(img, link).write
        } else
          value
      }
    }
  }

  class TaskFieldFormat(fieldName: String, displayName: Enum[_]) extends FieldFormat(fieldName, displayName) {

    def formatField(context: MailContext[_]): String = {
      val task = context.getWrappedBean.asInstanceOf[SimpleTask]
      if (task.getParenttaskid != null) {
        val img = new Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.TASK))
        val parentTaskLink = ProjectLinkGenerator.generateTaskPreviewFullLink(context.siteUrl, task.getParentTaskKey,
          task.getProjectShortname)
        val link = FormatUtils.newA(parentTaskLink, task.getName)
        FormatUtils.newLink(img, link).write
      }
      else {
        new Span().write
      }
    }

    def formatField(context: MailContext[_], value: String): String = {
      if (StringUtils.isBlank(value)) {
        return new Span().write
      }
      try {
        val taskId = value.toInt
        val taskService = AppContextUtil.getSpringBean(classOf[ProjectTaskService])
        val task = taskService.findById(taskId, context.getUser.getAccountId)
        if (task != null) {
          val img = new Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.TASK));
          val taskListLink = ProjectLinkGenerator.generateTaskPreviewFullLink(context.siteUrl, task.getTaskkey, task.getProjectShortname)
          val link = FormatUtils.newA(taskListLink, task.getName)
          return FormatUtils.newLink(img, link).write
        }
      }
      catch {
        case e: Exception => LOG.error("Error", e)
      }
      value
    }
  }

  class MilestoneFieldFormat(fieldName: String, displayName: Enum[_]) extends FieldFormat(fieldName, displayName) {

    def formatField(context: MailContext[_]): String = {
      val task = context.getWrappedBean.asInstanceOf[SimpleTask]
      if (task.getMilestoneid != null) {
        val img = new Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.MILESTONE))
        val milestoneLink = ProjectLinkGenerator.generateMilestonePreviewFullLink(context.siteUrl, task.getProjectid,
          task.getMilestoneid)
        val link = FormatUtils.newA(milestoneLink, task.getMilestoneName)
        FormatUtils.newLink(img, link).write
      }
      else {
        new Span().write
      }
    }

    def formatField(context: MailContext[_], value: String): String = {
      if (StringUtils.isBlank(value)) {
        return new Span().write
      }
      try {
        val milestoneId = value.toInt
        val milestoneService = AppContextUtil.getSpringBean(classOf[MilestoneService])
        val milestone = milestoneService.findById(milestoneId, context.getUser.getAccountId)
        if (milestone != null) {
          val img = new Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.MILESTONE))
          val milestoneLink = ProjectLinkGenerator.generateMilestonePreviewFullLink(context.siteUrl,
            milestone.getProjectid, milestone.getId)
          val link = FormatUtils.newA(milestoneLink, milestone.getName)
          return FormatUtils.newLink(img, link).write
        }
      }
      catch {
        case e: Exception => LOG.error("Error", e)
      }
      value
    }
  }

}

File: mycollab-scheduler/src/main/java/com/mycollab/module/project/schedule/email/service/ProjectTaskRelayEmailNotificationActionImpl.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.project.schedule.email.service

import com.hp.gagawa.java.elements.A
import com.hp.gagawa.java.elements.Span
import com.hp.gagawa.java.elements.Text
import com.mycollab.common.MonitorTypeConstants
import com.mycollab.common.NotificationType
import com.mycollab.common.i18n.GenericI18Enum
import com.mycollab.core.MyCollabException
import com.mycollab.core.utils.StringUtils
import com.mycollab.html.FormatUtils
import com.mycollab.html.LinkUtils
import com.mycollab.module.mail.MailUtils
import com.mycollab.module.project.ProjectLinkGenerator
import com.mycollab.module.project.ProjectResources
import com.mycollab.module.project.ProjectTypeConstants
import com.mycollab.module.project.domain.ProjectRelayEmailNotification
import com.mycollab.module.project.domain.SimpleTask
import com.mycollab.module.project.domain.Task
import com.mycollab.module.project.i18n.MilestoneI18nEnum
import com.mycollab.module.project.i18n.OptionI18nEnum.Priority
import com.mycollab.module.project.i18n.TaskI18nEnum
import com.mycollab.module.project.service.MilestoneService
import com.mycollab.module.project.service.ProjectNotificationSettingService
import com.mycollab.module.project.service.ProjectTaskService
import com.mycollab.module.user.AccountLinkGenerator
import com.mycollab.module.user.domain.SimpleUser
import com.mycollab.module.user.service.UserService
import com.mycollab.schedule.email.ItemFieldMapper
import com.mycollab.schedule.email.MailContext
import com.mycollab.schedule.email.format.DateFieldFormat
import com.mycollab.schedule.email.format.FieldFormat
import com.mycollab.schedule.email.format.I18nFieldFormat
import com.mycollab.schedule.email.project.ProjectTaskRelayEmailNotificationAction
import com.mycollab.spring.AppContextUtil
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class ProjectTaskRelayEmailNotificationActionImpl : SendMailToFollowersAction<SimpleTask>(), ProjectTaskRelayEmailNotificationAction {

    @Autowired private lateinit var projectTaskService: ProjectTaskService

    @Autowired private lateinit var projectNotificationService: ProjectNotificationSettingService

    private val mapper = TaskFieldNameMapper()

    override fun buildExtraTemplateVariables(context: MailContext<SimpleTask>) {
        val emailNotification = context.emailNotification

        val summary = "#${bean!!.taskkey} - ${bean!!.name}"
        val summaryLink = ProjectLinkGenerator.generateTaskPreviewFullLink(siteUrl, bean!!.taskkey, bean!!.projectShortname!!)

        val avatarId = if (projectMember != null) projectMember!!.memberAvatarId else ""
        val userAvatar = LinkUtils.newAvatar(avatarId)

        val makeChangeUser = "${userAvatar.write()} ${emailNotification.changeByUserFullName}"
        val actionEnum = when (emailNotification.action) {
            MonitorTypeConstants.CREATE_ACTION -> TaskI18nEnum.MAIL_CREATE_ITEM_HEADING
            MonitorTypeConstants.UPDATE_ACTION -> TaskI18nEnum.MAIL_UPDATE_ITEM_HEADING
            MonitorTypeConstants.ADD_COMMENT_ACTION -> TaskI18nEnum.MAIL_COMMENT_ITEM_HEADING
            else -> throw MyCollabException("Not support action ${emailNotification.action}")
        }

        contentGenerator.putVariable("projectName", bean!!.projectName!!)
        contentGenerator.putVariable("projectNotificationUrl", ProjectLinkGenerator.generateProjectSettingFullLink(siteUrl, bean!!.projectid))
        contentGenerator.putVariable("actionHeading", context.getMessage(actionEnum, makeChangeUser))
        contentGenerator.putVariable("name", summary)
        contentGenerator.putVariable("summaryLink", summaryLink)
    }

    override fun getBeanInContext(notification: ProjectRelayEmailNotification): SimpleTask? =
            projectTaskService.findById(notification.typeid.toInt(), notification.saccountid)

    override fun getItemName(): String = StringUtils.trim(bean!!.name, 100)

    override fun getProjectName(): String = bean!!.name

    override fun getCreateSubject(context: MailContext<SimpleTask>): String =
            context.getMessage(TaskI18nEnum.MAIL_CREATE_ITEM_SUBJECT, bean!!.projectName, context.changeByUserFullName, getItemName())

    override fun getCreateSubjectNotification(context: MailContext<SimpleTask>): String =
            context.getMessage(TaskI18nEnum.MAIL_CREATE_ITEM_SUBJECT, projectLink(), userLink(context), taskLink())

    override fun getUpdateSubject(context: MailContext<SimpleTask>): String =
            context.getMessage(TaskI18nEnum.MAIL_UPDATE_ITEM_SUBJECT, bean!!.projectName, context.changeByUserFullName, getItemName())

    override fun getUpdateSubjectNotification(context: MailContext<SimpleTask>): String =
            context.getMessage(TaskI18nEnum.MAIL_UPDATE_ITEM_SUBJECT, projectLink(), userLink(context), taskLink())

    override fun getCommentSubject(context: MailContext<SimpleTask>): String =
            context.getMessage(TaskI18nEnum.MAIL_COMMENT_ITEM_SUBJECT, bean!!.projectName, context.changeByUserFullName, getItemName())

    override fun getCommentSubjectNotification(context: MailContext<SimpleTask>): String =
            context.getMessage(TaskI18nEnum.MAIL_COMMENT_ITEM_SUBJECT, projectLink(), userLink(context), taskLink())

    private fun projectLink() = A(ProjectLinkGenerator.generateProjectLink(bean!!.projectid)).appendText(bean!!.projectName).write()

    private fun userLink(context: MailContext<SimpleTask>) = A(AccountLinkGenerator.generateUserLink(context.user.username)).appendText(context.changeByUserFullName).write()

    private fun taskLink() = A(ProjectLinkGenerator.generateTaskPreviewLink(bean!!.taskkey, bean!!.projectShortname!!)).appendText(getItemName()).write()

    override fun getItemFieldMapper(): ItemFieldMapper = mapper

    override fun getListNotifyUsersWithFilter(notification: ProjectRelayEmailNotification): List<SimpleUser> {
        val notificationSettings = projectNotificationService.findNotifications(notification.projectId, notification.saccountid)
        var notifyUsers = notification.notifyUsers

        notificationSettings.forEach { notificationSetting ->
            if (NotificationType.None.name == notificationSetting.level) {
                notifyUsers = notifyUsers.filter { it.username != notificationSetting.username }
            } else if (NotificationType.Minimal.name == notificationSetting.level) {
                val findResult = notifyUsers.find { it.username == notificationSetting.username }
                if (findResult == null) {
                    val task = projectTaskService.findById(notification.typeid.toInt(), notification.saccountid)
                    if (task != null && notificationSetting.username == task.assignuser) {
                        val prjMember = projectMemberService.getActiveUserOfProject(notificationSetting.username,
                                notificationSetting.projectid, notificationSetting.saccountid)
                        if (prjMember != null) {
                            notifyUsers += prjMember
                        }
                    }
                }
            } else if (NotificationType.Full.name == notificationSetting.level) {
                val prjMember = projectMemberService.getActiveUserOfProject(notificationSetting.username,
                        notificationSetting.projectid, notificationSetting.saccountid)
                if (prjMember != null) {
                    notifyUsers += prjMember
                }
            }
        }

        return notifyUsers
    }

    override fun getType(): String = ProjectTypeConstants.TASK

    override fun getTypeId(): String = "${bean!!.id}"

    class TaskFieldNameMapper : ItemFieldMapper() {
        init {
            put(Task.Field.name, GenericI18Enum.FORM_NAME, isColSpan = true)
            put(Task.Field.startdate, DateFieldFormat(Task.Field.startdate.name, GenericI18Enum.FORM_START_DATE))
            put(Task.Field.enddate, DateFieldFormat(Task.Field.enddate.name, GenericI18Enum.FORM_END_DATE))
            put(Task.Field.duedate, DateFieldFormat(Task.Field.duedate.name, GenericI18Enum.FORM_DUE_DATE))
            put(Task.Field.percentagecomplete, TaskI18nEnum.FORM_PERCENTAGE_COMPLETE)
            put(Task.Field.priority, I18nFieldFormat(Task.Field.priority.name, GenericI18Enum.FORM_PRIORITY, Priority::class.java))
            put(Task.Field.assignuser, AssigneeFieldFormat(Task.Field.assignuser.name, GenericI18Enum.FORM_ASSIGNEE))
            put(Task.Field.isestimated, TaskI18nEnum.FORM_IS_ESTIMATED)
            put(Task.Field.remainestimate, TaskI18nEnum.FORM_REMAIN_ESTIMATE)
            put(Task.Field.milestoneid, MilestoneFieldFormat(Task.Field.milestoneid.name, MilestoneI18nEnum.SINGLE))
            put(Task.Field.parenttaskid, TaskFieldFormat(Task.Field.parenttaskid.name, TaskI18nEnum.FORM_PARENT_TASK))
            put(Task.Field.description, GenericI18Enum.FORM_DESCRIPTION)
            put(Task.Field.status, GenericI18Enum.FORM_STATUS)
        }
    }

    class AssigneeFieldFormat(fieldName: String, displayName: Enum<*>) : FieldFormat(fieldName, displayName) {

        override fun formatField(context: MailContext<*>): String {
            val task = context.wrappedBean as SimpleTask
            return if (task.assignuser != null) {
                val userAvatarLink = MailUtils.getAvatarLink(task.assignUserAvatarId, 16)
                val img = FormatUtils.newImg("avatar", userAvatarLink)
                val userLink = AccountLinkGenerator.generatePreviewFullUserLink(MailUtils.getSiteUrl(task.saccountid), task.assignuser)
                val link = FormatUtils.newA(userLink, task.assignUserFullName)
                FormatUtils.newLink(img, link).write()
            } else Span().write()
        }

        override fun formatField(context: MailContext<*>, value: String): String {
            return if (StringUtils.isBlank(value)) {
                Span().write()
            } else {
                val userService = AppContextUtil.getSpringBean(UserService::class.java)
                val user = userService.findUserByUserNameInAccount(value, context.saccountid)
                if (user != null) {
                    val userAvatarLink = MailUtils.getAvatarLink(user.avatarid, 16)
                    val userLink = AccountLinkGenerator.generatePreviewFullUserLink(MailUtils.getSiteUrl(context.saccountid), user.username)
                    val img = FormatUtils.newImg("avatar", userAvatarLink)
                    val link = FormatUtils.newA(userLink, user.displayName!!)
                    FormatUtils.newLink(img, link).write()
                } else value
            }
        }
    }

    class TaskFieldFormat(fieldName: String, displayName: Enum<*>) : FieldFormat(fieldName, displayName) {

        override fun formatField(context: MailContext<*>): String {
            val task = context.wrappedBean as SimpleTask
            return if (task.parenttaskid != null) {
                val img = Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.TASK))
                val parentTaskLink = ProjectLinkGenerator.generateTaskPreviewFullLink(context.siteUrl, task.parentTaskKey,
                        task.projectShortname!!)
                val link = FormatUtils.newA(parentTaskLink, task.name)
                FormatUtils.newLink(img, link).write()
            } else Span().write()
        }

        override fun formatField(context: MailContext<*>, value: String): String {
            if (StringUtils.isBlank(value)) {
                return Span().write()
            }

            val taskId = value.toInt()
            val taskService = AppContextUtil.getSpringBean(ProjectTaskService::class.java)
            val task = taskService.findById(taskId, context.saccountid)
            return if (task != null) {
                val img = Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.TASK))
                val taskListLink = ProjectLinkGenerator.generateTaskPreviewFullLink(context.siteUrl, task.taskkey, task.projectShortname!!)
                val link = FormatUtils.newA(taskListLink, task.name)
                return FormatUtils.newLink(img, link).write()
            } else value
        }
    }

    class MilestoneFieldFormat(fieldName: String, displayName: Enum<*>) : FieldFormat(fieldName, displayName) {

        override fun formatField(context: MailContext<*>): String {
            val task = context.wrappedBean as SimpleTask
            return if (task.milestoneid != null) {
                val img = Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.MILESTONE))
                val milestoneLink = ProjectLinkGenerator.generateMilestonePreviewFullLink(context.siteUrl, task.projectid,
                        task.milestoneid)
                val link = FormatUtils.newA(milestoneLink, task.milestoneName!!)
                FormatUtils.newLink(img, link).write()
            } else Span().write()
        }

        override fun formatField(context: MailContext<*>, value: String): String {
            if (StringUtils.isBlank(value)) {
                return Span().write()
            }

            val milestoneId = value.toInt()
            val milestoneService = AppContextUtil.getSpringBean(MilestoneService::class.java)
            val milestone = milestoneService.findById(milestoneId, context.saccountid)
            return if (milestone != null) {
                val img = Text(ProjectResources.getFontIconHtml(ProjectTypeConstants.MILESTONE))
                val milestoneLink = ProjectLinkGenerator.generateMilestonePreviewFullLink(context.siteUrl,
                        milestone.projectid, milestone.id)
                val link = FormatUtils.newA(milestoneLink, milestone.name)
                FormatUtils.newLink(img, link).write()
            } else value
        }
    }

}